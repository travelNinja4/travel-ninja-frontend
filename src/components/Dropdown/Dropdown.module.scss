/**
------------------------------------------------------
styles specific to the Dropdown component
------------------------------------------------------
**/

.dropdown {
  position: relative;
}

.inputWrapper {
  position: relative;
  display: flex;
  flex-direction: column;
}

.label {
  font-weight: 600;
  font-size: 0.95rem;
  color: var(--color-gray-800);
  margin-bottom: 0.5rem;
}

.input {
  flex: 1;
  width: 100%;
  font-size: 1rem;
  padding: 0.75rem 0.87rem;
  border-radius: 8px;
  border: 1px solid var(--color-gray-300);
  background-color: var(--color-white);
  color: var(--color-black);
  color-scheme: light;
  box-sizing: border-box;
  transition:
    border-color 0.2s ease,
    box-shadow 0.2s ease;

  &:focus {
    outline: none;
    border-color: var(--color-indigo);
    box-shadow: 0 0 0 0.2rem var(--color-indigo-alpha-25);
    outline: none;
  }

  &:disabled {
    background-color: #e9ecef;
    color: var(--color-gray-400);
    cursor: not-allowed;
  }
}

.list {
  position: absolute;
  top: calc(100% + 0.25rem);
  left: 0;
  width: 100%;
  margin: 0;
  padding: 0;
  list-style: none;
  border: 1px solid var(--color-gray-220);
  border-radius: 6px;
  background: var(--color-white);
  box-shadow: 0 4px 8px var(--color-black-alpha-10);
  max-height: 12rem;
  overflow-y: auto;
  z-index: 100;

  // smooth open/close animation
  animation: fadeIn 0.15s ease-in-out;
}

.option {
  padding: 0.5rem 0.75rem;
  cursor: pointer;
  font-size: 0.95rem;
  color: var(--color-gray-700);
  transition:
    background 0.2s ease,
    color 0.2s ease;

  &:hover,
  &.highlighted {
    background: var(--color-option-hover-bg);
  }

  &.selected {
    font-weight: 600;
    background: darker(var(--color-option-selected-bg), 15%);
    color: var(--color-option-selected-text);
  }
}

.noOptions {
  padding: 0.5rem 0.75rem;
  font-size: 0.9rem;
  color: var(--color-gray-350);
  text-align: center;
}

.inputWithIcon {
  position: relative;
  display: flex;
  align-items: center;

  input {
    width: 100%;
    padding-right: 2.5rem;
  }
}

.chevronIcon {
  position: absolute;
  right: 0.75rem;
  top: 50%;
  transform: translateY(-50%);
  pointer-events: none;
  color: #000000;
  transition: transform 0.2s ease;
}

.loading {
  display: flex;
  align-items: center;
  gap: 0.5rem;
  color: var(--color-gray-320);
  font-size: 0.875rem;
  height: 45px;
  .spinner {
    height: 1rem;
    width: 1rem;
    border: 2px solid var(--icon-color-info);
    border-top-color: transparent;
    border-radius: 9999px;
    animation: spin 1s linear infinite;
  }
}

@keyframes spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}

@keyframes fadeIn {
  from {
    opacity: 0;
    transform: translateY(-4px);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
}
