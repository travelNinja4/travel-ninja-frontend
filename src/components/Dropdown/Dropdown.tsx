/**
 * A reusable dropdown component that allows users to select an option from a list. It supports customizable labels, dynamic options, and flexible styling for use across the application.
 *
 * @example
 * ```tsx
 * import Dropdown from '@src/components/Dropdown'
 *
 * export default function Dropdown() {
 *   return <Dropdown label="Hello" />;
 * }
 * ```
 */
'use client';

import { useState, useRef, useEffect, KeyboardEvent, ChangeEvent } from 'react';
import { STRINGS } from '@/constants/strings';
import Typography from '../Typography';
import TextField from '../TextField';
import { ChevronDown, ChevronUp } from 'lucide-react';
import clsx from 'clsx';
import styles from './Dropdown.module.scss';

export type DropdownOption<T> = {
  value: T;
  label: string;
};

/**
 * Define the props available for the Dropdown component.
 */
interface DropdownProps<T> {
  /**
   * The label displayed above the dropdown input.
   */
  label?: string;

  /**
   * If true, indicates that selecting a value is required.
   * Typically adds a visual indicator (like a *) next to the label.
   */
  required?: boolean;

  /**
   * List of options to display in the dropdown.
   *
   * Each option must have a `label` (string shown to the user)
   * and a `value` (the underlying data of generic type `T`).
   */
  options: DropdownOption<T>[];

  /**
   * The currently selected value.
   *
   * Can be a string or of type `T`. Useful for controlled components
   * where the parent manages the selected state.
   */
  value?: T | string;

  /**
   * Callback fired when an option is selected.
   *
   * @param value The selected value of type `T` or string.
   * If cleared, it may be `undefined`.
   */
  onChange?: (value: T | string | undefined) => void;

  /**
   * Placeholder text for the search input field.
   *
   * @default "Search..."
   */
  placeholder?: string;

  /**
   * Additional class name(s) applied to the dropdown container.
   * Allows custom styling.
   */
  dropdownContainerClassName?: string;

  /**
   * Additional class name(s) applied to the text input container.
   * Allows custom styling.
   */
  dropdownTextContainerClassName?: string;

  /**
   * If true, disables the input and dropdown interaction.
   *
   * @default false
   */
  disabled?: boolean;

  /**
   * If true, shows loading state in the dropdown options.
   */
  isLoading?: boolean;
}

export default function Dropdown<T>({
  label,
  required,
  options,
  value,
  onChange,
  placeholder = 'Select...',
  dropdownContainerClassName,
  dropdownTextContainerClassName,
  disabled = false,
  isLoading,
}: DropdownProps<T>) {
  const [highlightedIndex, setHighlightedIndex] = useState<number>(0);
  const [isOpen, setIsOpen] = useState(false);
  const dropdownRef = useRef<HTMLDivElement>(null);

  // Close dropdown when clicking outside
  useEffect(() => {
    const handleClickOutside = (e: MouseEvent) => {
      if (dropdownRef.current && !dropdownRef.current.contains(e.target as Node)) {
        setIsOpen(false);
      }
    };
    document.addEventListener('mousedown', handleClickOutside);
    return () => document.removeEventListener('mousedown', handleClickOutside);
  }, []);

  // Keyboard navigation
  const handleKeyDown = (e: KeyboardEvent<HTMLInputElement | HTMLUListElement>) => {
    if (!isOpen) return;

    if (e.key === 'ArrowDown') {
      setHighlightedIndex((prev) => (prev + 1) % options.length);
      e.preventDefault();
    } else if (e.key === 'ArrowUp') {
      setHighlightedIndex((prev) => (prev - 1 + options.length) % options.length);
      e.preventDefault();
    } else if (e.key === 'Enter') {
      const option = options[highlightedIndex];
      onChange?.(option.value);
      setIsOpen(false);
      e.preventDefault();
    } else if (e.key === 'Escape') {
      setIsOpen(false);
      e.preventDefault();
    }
  };

  return (
    <div
      data-testid="DropdownTest"
      ref={dropdownRef}
      tabIndex={0}
      className={clsx(styles.dropdown, dropdownContainerClassName)}
    >
      <div className={styles.inputWrapper}>
        {label && (
          <Typography tag="label" className={styles.label}>
            {label}
            {required && (
              <Typography tag="span" className={styles.requiredMark}>
                *
              </Typography>
            )}
          </Typography>
        )}

        <div
          className={styles.inputWithIcon}
          onClick={() => !disabled && setIsOpen((prev) => !prev)}
        >
          <TextField
            readOnly
            placeholder={placeholder}
            value={value as string}
            className={clsx(dropdownTextContainerClassName)}
            onFocus={() => !disabled && setIsOpen(true)}
            onKeyDown={handleKeyDown}
            disabled={disabled}
          />

          {/* Chevron inside the input */}
          <span className={clsx(styles.chevronIcon)}>
            {isOpen ? <ChevronUp size={16} /> : <ChevronDown size={16} />}
          </span>
        </div>
      </div>

      {isOpen && (
        <ul role="listbox" tabIndex={-1} onKeyDown={handleKeyDown} className={styles.list}>
          {isLoading ? (
            // Loading State (fixed)
            <li className={styles.loading}>
              <span className={styles.spinner}></span>
              Loading...
            </li>
          ) : options?.length > 0 ? (
            //  Normal Options
            options.map((option, idx) => {
              const isHighlighted = idx === highlightedIndex;
              return (
                <li
                  key={idx}
                  role="option"
                  className={clsx(styles.option, {
                    [styles.highlighted]: isHighlighted,
                    [styles.selected]: option.value === value,
                  })}
                  onMouseEnter={() => setHighlightedIndex(idx)}
                  onClick={() => {
                    onChange?.(option.value);
                    setIsOpen(false);
                  }}
                >
                  {option.label}
                </li>
              );
            })
          ) : (
            // No Options Fallback
            <li className={styles.noOptions}>{STRINGS.NO_OPTIONS_FOUND}</li>
          )}
        </ul>
      )}
    </div>
  );
}
