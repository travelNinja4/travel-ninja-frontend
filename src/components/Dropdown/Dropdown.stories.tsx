import type { Meta, StoryObj } from '@storybook/nextjs-vite';
/**
import { fn } from 'storybook/test';
*/
import Dropdown from './Dropdown'

const meta = {
  title: 'components/Dropdown',
  component: Dropdown,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    label: { 
        control: 'text', 
        description: 'Sample label for the component' 
        },
  },
  args: { 
    /** onClick: fn() **/
    },
} satisfies Meta<typeof Dropdown>;

export default meta;
type Story = StoryObj<typeof meta>;