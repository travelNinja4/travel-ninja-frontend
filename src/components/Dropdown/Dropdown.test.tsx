import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import Dropdown from './Dropdown'

describe('Home', () => {
  /** Base props for Dropdown **/
  const baseProps={
    label: 'Dropdown Test'
  }
 
  /**
  * @test
  * Renders the Dropdown component without throwing an error
  * and checks if the DropdownTest text is present in the document.
  *
  * This ensures that the component loads correctly with the base props.
  */
  it('renders without crashing', () => {
   render(<Dropdown {...baseProps}/>);
   expect(screen.getByTestId('DropdownTest')).toBeInTheDocument();
  });

  /**
  * @test
  * Ensures that the Dropdown component renders without throwing any runtime errors.
  *
  * This is a safety check to confirm that the component can mount successfully
  * with the provided base props, even if it does not render any visible content.
  */
  it('renders without throwing', () => {
  expect(() => render(<Dropdown {...baseProps}/>)).not.toThrow();
  });

  /**
  * @test
  * Captures a snapshot of the Dropdown component with base props
  * and ensures it matches the previously stored snapshot.
  *
  * This helps detect any unexpected UI changes.
  */
  it('matches snapshot', () => {
   const { asFragment } = render(<Dropdown {...baseProps}/>);
   expect(asFragment()).toMatchSnapshot();
  });
})