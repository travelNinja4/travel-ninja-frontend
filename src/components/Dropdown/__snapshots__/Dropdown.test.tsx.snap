// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Home matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="dropdown"
    data-testid="DropdownTest"
    tabindex="0"
  >
    <div
      class="inputWrapper"
    >
      <label
        class="typography label align-left label"
        data-testid="TypographyTest"
      >
        Dropdown Test
      </label>
      <div
        class="inputWithIcon"
      >
        <div
          class="container"
          data-testid="TextFieldTest"
        >
          <div
            class="inputWrapper"
          >
            <input
              class="inputField"
              placeholder="Select..."
              readonly=""
              type="text"
            />
          </div>
        </div>
        <span
          class="chevronIcon"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down"
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
